compiler/compiler.go

```diff
@@ -40,24 +46,80 @@ type ast struct {
 // FileDescriptorSet is the intermediary representation typically
 // passed to proto plugins.
 func NewFileDescriptorSet(files, importPaths []string, includeImports bool) (*pb.FileDescriptorSet, error) {
+	all, filtered, err := compileFileDescriptorSets(files, importPaths, includeImports)
+	if err != nil {
+		return nil, err
+	}
+	err = resolveCustomOptions(all, filtered)
+	if err != nil {
+		return nil, err
+	}
+	return filtered, nil
+}
+
+func compileFileDescriptorSets(files, importPaths []string, includeImports bool) (all *pb.FileDescriptorSet, filtered *pb.FileDescriptorSet, err error) {
 	done := map[string]bool{}
 	origFiles := map[string]bool{}
 	for _, file := range files {
 		origFiles[file] = true
 	}
 	asts, err := readProtos(files, importPaths, done)
 	if err != nil {
-		return nil, err
+		return nil, nil, err
 	}
 	types := newTypes(asts)
-	fds := &pb.FileDescriptorSet{}
+	all = &pb.FileDescriptorSet{}
+	filtered = &pb.FileDescriptorSet{}
 	for _, a := range asts {
+		fd := newFileDescriptor(a, types)
+		all.File = append(all.File, fd)
 		if includeImports || origFiles[a.file] {
-			fd := newFileDescriptor(a, types)
-			fds.File = append(fds.File, fd)
+			filtered.File = append(filtered.File, fd)
+		}
+	}
+	return all, filtered, nil
+}
+
+func resolveCustomOptions(all, filtered *pb.FileDescriptorSet) error {
```

Comment:
What's filtered?

Reply:
if `includeImports` is set to false, filtered only includes the files explicitly listed and not their (transitive) dependencies.

This analogous to `protoc` without  `--include_imports`.
However, we need to keep track of the filedescriptors of all dependencies in order to resolve the extensions of custom options.

I will add a comment. Do you have any suggestions for a better name?